File: backend/app/schemas/recording.py
from typing import List, Optional
from datetime import datetime
from pydantic import BaseModel, ConfigDict
from app.schemas.analytics import ScoreTrendPoint


class PracticeRecordingResponse(BaseModel):
    id: int
    title: str
    mime_type: str
    duration_seconds: Optional[int] = None
    file_size_bytes: int
    interview_question_id: Optional[int] = None
    created_at: datetime

    model_config = ConfigDict(from_attributes=True)


class RecordingCommunicationScore(BaseModel):
    category: str
    score: float
    max_score: float = 10.0
    feedback: str


class RecordingAnalysisResponse(BaseModel):
    id: int
    recording_id: int
    provider: Optional[str] = None
    transcript: Optional[str] = None
    communication_scores: List[RecordingCommunicationScore] = []
    filler_word_count: Optional[int] = None
    summary: Optional[str] = None
    content_scores: List[RecordingCommunicationScore] = []
    content_summary: Optional[str] = None
    analysis_status: str
    analysis_error: Optional[str] = None
    created_at: datetime

    model_config = ConfigDict(from_attributes=True)


class AnalyzeRecordingRequest(BaseModel):
    provider: Optional[str] = None


class ProviderInfo(BaseModel):
    name: str
    is_available: bool


class RoundAnalyticsItem(BaseModel):
    round_type: str
    round_label: str
    attempt_count: int
    avg_content_score_pct: Optional[float] = None
    avg_delivery_score_pct: Optional[float] = None


class WeakestContentCategory(BaseModel):
    category: str
    round_label: str
    avg_score_pct: float


class RecordingAnalytics(BaseModel):
    total_recordings: int
    analyzed_count: int
    by_round: List[RoundAnalyticsItem] = []
    delivery_trend: List[ScoreTrendPoint] = []
    weakest_content_category: Optional[WeakestContentCategory] = None

File: frontend/src/types/recording.ts
// PrepBench - Copyright (c) 2026 Nimish Kanungo
// Licensed under the PolyForm Noncommercial License 1.0.0 (see LICENSE).
// Commercial use requires a separate licence from the copyright holder.

import { ScoreTrendPoint } from './analytics';

export interface PracticeRecording {
  id: number;
  title: string;
  mime_type: string;
  duration_seconds: number | null;
  file_size_bytes: number;
  interview_question_id: number | null;
  session_id?: number | null;
  plan_note?: string | null;
  created_at: string;
  /** List endpoint only. Percentages stay null unless the analysis succeeded. */
  analysis_status?: string | null;
  content_percent?: number | null;
  delivery_percent?: number | null;
}

export interface RecordingCommunicationScore {
  category: string;
  score: number;
  max_score: number;
  feedback: string;
}

export interface KeyPointMatch {
  point: string;
  status: 'covered' | 'partial' | 'missed';
  evidence: string;
}

export interface AnswerComparison {
  alignment_score: number;
  key_point_matches: KeyPointMatch[];
  gap_analysis?: string | null;
  unplanned_additions?: string | null;
  coaching_tips?: string | null;
}

export type AnalysisStatus = 'analyzed' | 'unavailable' | 'error';

export interface RecordingAnalysis {
  id: number;
  recording_id: number;
  provider: string | null;
  transcript: string | null;
  communication_scores: RecordingCommunicationScore[];
  filler_word_count: number | null;
  summary: string | null;
  content_scores: RecordingCommunicationScore[];
  content_summary: string | null;
  answer_comparison?: AnswerComparison | null;
  analysis_status: AnalysisStatus;
  analysis_error: string | null;
  created_at: string;
}

export interface ProviderInfo {
  name: string;
  is_available: boolean;
}

export interface RoundAnalyticsItem {
  round_type: string;
  round_label: string;
  attempt_count: number;
  avg_content_score_pct: number | null;
  avg_delivery_score_pct: number | null;
}

export interface WeakestContentCategory {
  category: string;
  round_label: string;
  avg_score_pct: number;
}

export interface RecordingAnalytics {
  total_recordings: number;
  analyzed_count: number;
  by_round: RoundAnalyticsItem[];
  delivery_trend: ScoreTrendPoint[];
  weakest_content_category: WeakestContentCategory | null;
}
